scripts/lsp_types.lua

```diff
@@ -18,19 +19,90 @@ local function tofile(fname, text)
   end
 end
 
-function M.gen(opt)
-  if vim.uv.fs_stat('./lsp.json') then
-    vim.fn.delete('./lsp.json')
+local function read_json(opt)
+  local uri = 'https://raw.githubusercontent.com/microsoft/language-server-protocol/gh-pages/_specifications/lsp/'
+    .. opt.version
+    .. '/metaModel/metaModel.json'
+
+  local res = vim
+    .system({
+      'curl',
+      uri,
+      '-o',
+      '-',
+    })
+    :wait()
+
+  if res.code ~= 0 then
+    print(res.stderr)
+    return
   end
-  vim.fn.system({
-    'curl',
-    'https://raw.githubusercontent.com/microsoft/lsprotocol/' .. opt.ref .. '/generator/lsp.json',
-    '-o',
-    './lsp.json',
+  return vim.json.decode(res.stdout)
+end
+
+local function gen_methods(protocol)
+  local output =
+    { '--- LSP methods. Generated by lsp_types.lua, keep at end of file.', 'protocol.Methods = {' }
+  local indent = (' '):rep(2)
+
+  vim.iter(protocol.requests):map(function(item)
+    if item.method then
+      local name = item.method:gsub('/', '_', 3)
+      if item.documentation then
+        local document = vim.split(item.documentation, '\n?\n', { trimempty = true })
+        vim.iter(document):map(function(docstring)
+          output[#output + 1] = indent .. '--- ' .. docstring
```

Comment:
why is iter/map being used here instead of ipairs()? Always use simple, common idioms unless there is an actual reason that a higher level abstraction is needed.